share/shwap/row.go

```diff
@@ -0,0 +1,139 @@
+package shwap
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/pkg/wrapper"
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+	types_pb "github.com/celestiaorg/celestia-node/share/shwap/proto"
+)
+
+// RowSide enumerates the possible sides of a row within an Extended Data Square (EDS).
+type RowSide int
+
+const (
+	Left  RowSide = iota // Left side of the row.
+	Right                // Right side of the row.
+)
+
+// Row represents a portion of a row in an EDS, either left or right half.
+type Row struct {
+	halfShares []share.Share // halfShares holds the shares of either the left or right half of a row.
+	side       RowSide       // side indicates whether the row half is left or right.
+}
+
+// NewRow creates a new Row with the specified shares and side.
+func NewRow(halfShares []share.Share, side RowSide) Row {
+	return Row{
+		halfShares: halfShares,
+		side:       side,
+	}
+}
+
+// Validate checks if the row's shares match the expected number from the root data and validates
+// the side of the row.
+func (r Row) Validate(dah *share.Root, idx int) error {
+	if len(r.halfShares) == 0 {
+		return fmt.Errorf("empty half row")
+	}
+	expectedShares := len(dah.RowRoots) / 2
+	if len(r.halfShares) != expectedShares {
+		return fmt.Errorf("shares size doesn't match root size: %d != %d", len(r.halfShares), expectedShares)
+	}
+	if r.side != Left && r.side != Right {
```

Comment:
This condition is necessary to validate input received from the network. We need to evaluate it to prevent unexpected behavior if the received message has a side value of 6 for example.

Reply:
I see, given that, I'd almost like to see a `r.Valid()` or `r.SideWithinBounds()` or something descriptive to make what we're checking super clear and literal. Not essential, but I didn't realize we could get noise from network, so it feels some checking/validation should occur in a nicely encapsulated way